File: src/SearchInsertPosition.java
/**
 * Created by kaijiezhou on 7/3/15.
 */
public class SearchInsertPosition {
    public int searchInsert(int[] nums, int target) {
        if(nums==null&&nums.length==0) return 0;
        int index=binarySearch(nums,target,0,nums.length-1);
        if(nums[index]>target) return index+1;
        else return index;
    }
    private int binarySearch(int[]nums, int target, int start, int end){
        int mid=(start+end)/2;
        if(nums[mid]==target){
            return mid;
        }else if(mid==start&&mid==end){
            return mid;
        }else if(nums[mid]>target){
            return binarySearch(nums,target,start,mid);
        }else{
            return binarySearch(nums,target,mid+1,end);
        }
    }
}

File: python/search_insert_position.py
__author__ = 'kaijiezhou'


class Solution(object):
    def searchInsert(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        position=self.binary_search(nums,target, 0, len(nums)-1)
        if nums[position]<target:
            return position+1
        return position

    def binary_search(self, nums, target, start, end):
        mid=(start+end)/2
        if nums[mid]==target:
            return mid
        elif end<=start:
            return start
        elif nums[mid]>target:
            return self.binary_search(nums, target, start, mid-1)
        else:
            return self.binary_search(nums, target, mid+1, end)

if __name__ == '__main__':
    print Solution().searchInsert([1,3],4)